pkg/controllers/vrep/replication_status_reconciler.go

```diff
@@ -0,0 +1,279 @@
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package vrep
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	vapi "github.com/vertica/vertica-kubernetes/api/v1"
+	"github.com/vertica/vertica-kubernetes/api/v1beta1"
+	"github.com/vertica/vertica-kubernetes/pkg/cmds"
+	"github.com/vertica/vertica-kubernetes/pkg/controllers"
+	verrors "github.com/vertica/vertica-kubernetes/pkg/errors"
+	"github.com/vertica/vertica-kubernetes/pkg/events"
+	vmeta "github.com/vertica/vertica-kubernetes/pkg/meta"
+	"github.com/vertica/vertica-kubernetes/pkg/names"
+	"github.com/vertica/vertica-kubernetes/pkg/podfacts"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin/opts/replicationstatus"
+	"github.com/vertica/vertica-kubernetes/pkg/vk8s"
+	vrepstatus "github.com/vertica/vertica-kubernetes/pkg/vrepstatus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	statusStarted   = "started"
+	statusFailed    = "failed"
+	statusCompleted = "completed"
+
+	opLoadSnapshotPrep = "load_snapshot_prep"
+	opDataTransfer     = "data_transfer"
+	opLoadSnapshot     = "load_snapshot"
+)
+
+type ReplicationStatusReconciler struct {
+	client.Client
+	VRec         *VerticaReplicatorReconciler
+	Vrep         *v1beta1.VerticaReplicator
+	dispatcher   vadmin.Dispatcher
+	SourcePFacts *podfacts.PodFacts
+	TargetPFacts *podfacts.PodFacts
+	Log          logr.Logger
+	TargetInfo   *ReplicationInfo
+}
+
+func MakeReplicationStatusReconciler(cli client.Client, r *VerticaReplicatorReconciler, vrep *v1beta1.VerticaReplicator,
+	log logr.Logger) controllers.ReconcileActor {
+	return &ReplicationStatusReconciler{
+		Client:     cli,
+		VRec:       r,
+		Vrep:       vrep,
+		Log:        log.WithName("ReplicationStatusReconciler"),
+		TargetInfo: &ReplicationInfo{},
+	}
+}
+
+func (r *ReplicationStatusReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	// no-op if replication is not being done asynchronously
+	if r.Vrep.Spec.Mode != replicationModeAsync {
+		r.Log.Info("Stopped reconciling status: not async")
+		return ctrl.Result{}, nil
+	}
+
+	// no-op if ReplicationComplete is present (either true or false)
+	isPresent := r.Vrep.IsStatusConditionPresent(v1beta1.ReplicationComplete)
+	if isPresent {
+		r.Log.Info("Stopped reconciling status: complete")
+		return ctrl.Result{}, nil
+	}
+
+	// no-op if ReplicationReady is false
+	isFalse := r.Vrep.IsStatusConditionFalse(v1beta1.ReplicationReady)
+	if isFalse {
+		r.Log.Info("Stopped reconciling status: not ready")
+		return ctrl.Result{}, nil
+	}
+
+	// no-op if there is no transaction ID we can use to query replication status
+	if r.Vrep.Status.TransactionID == 0 {
+		r.Log.Info("Stopped reconciling status: transaction ID 0")
```

Comment:
We should add one more check to see if replicating is started, if it is, we should return an error saying "cannot find transcation ID".